controller/share_manager_controller.go

```diff
@@ -0,0 +1,626 @@
+	if volume.Status.KubernetesStatus.LastPodRefAt != "" || volume.Status.KubernetesStatus.LastPVCRefAt != "" {
+		isVolumeAttached := volume.Status.State == types.VolumeStateAttached || volume.Status.State == types.VolumeStateAttaching
+		if sm.Status.State != types.ShareManagerStateStopped && isVolumeAttached && !volume.Status.FrontendDisabled {
+			log.WithField("volume", volume.Name).Infof("Stopping share manager, requesting Volume detach from node %v", volume.Spec.NodeID)
+			volume.Spec.NodeID = ""
+			if volume, err = c.ds.UpdateVolume(volume); err != nil {
+				return err
+			}
+		}
+		sm.Status.State = types.ShareManagerStateStopped
+		return c.cleanupShareManagerPod(sm)
+	}
+
+	if pod == nil {
+		if pod, err = c.createShareManagerPod(sm); err != nil {
+			sm.Status.State = types.ShareManagerStateError
+			log.WithError(err).Error("failed to create pod for share manager")
+			return err
+		}
+		sm.Status.State = types.ShareManagerStateStarting
+	}
+
+	switch pod.Status.Phase {
+	case v1.PodPending:
+		if sm.Status.State != types.ShareManagerStateStarting {
+			sm.Status.State = types.ShareManagerStateError
+			log.Errorf("Share Manager has state %v but the related pod is pending.", sm.Status.State)
+		}
+	case v1.PodRunning:
+		// pod readiness is based on the availability of the nfs server
+		// nfs server is only started after the volume is attached and mounted
+		allContainersReady := true
+		for _, st := range pod.Status.ContainerStatuses {
+			allContainersReady = allContainersReady && st.Ready
+		}
+
+		if !allContainersReady {
+			// ensure that the shared volume is attached to this share managers node, if the share manager
+			// switches nodes this will ensure, that the volumes will travel with the manager
+			if volume.Spec.NodeID != sm.Status.OwnerID {
+				// HACK: at the moment the consumer needs to control the state transitions of the volume
+				//  since it's not possible to just specify the desired state, we should fix that down the line.
+				//  for the actual state transition we need to request detachment then wait for detachment
+				//  afterwards we can request attachment to the new node.
+				if volume.Status.State == types.VolumeStateAttached || volume.Status.State == types.VolumeStateAttaching {
+					log.WithField("volume", volume.Name).Infof("Requesting Volume detach from previous node %v", volume.Spec.NodeID)
+					volume.Spec.NodeID = ""
+				}
+
+				if volume.Status.State == types.VolumeStateDetached {
+					log.WithField("volume", volume.Name).Info("Requesting Volume attach to share manager node")
+					volume.Spec.NodeID = sm.Status.OwnerID
+				}
+
+				if volume, err = c.ds.UpdateVolume(volume); err != nil {
+					return err
+				}
+			} else {
+				// manually requeue the share manager so we can check again for container ready just in case
+				key := volume.Namespace + "/" + volume.Name
+				c.queue.AddRateLimited(key)
+			}
+		} else if sm.Status.State == types.ShareManagerStateStarting {
```

Comment:
Can the share manager be directly updated from state stopped to state running?

Reply:
- If there is no pod we create one and transition into "starting" state

- If there is a pending pod but we are not in "starting" state, we transition to "error" state
- If there is a running pod but we are not in "starting / running" state we transition to "error" state

- If we are in "error" state we delete the pod

So no it shouldn't be able to go from stopped into running state.